### ISSUE TITLE
SPDK 20.04.1 - Enabling CRYPTO results in compile errors CentOS 7.5 gcc 4.8.5

### ISSUE THREAD

**Issue by nufosmatic** [spdk/spdk#1512]
SPDK 20.04.1 is otherwise working (nvme, bdev, aio, fio_plugin, perf). Enabled CRYPTO=y and now I'm getting compilation errors.

## Expected Behavior

Should just compile.

## Current Behavior

Lotso compile errors.

`In file included from sse/snow3g_sse.c:42:0:
./include/snow3g_common.h: In function 'C0_C11_4':
./include/snow3g_common.h:125:9: warning: implicit declaration of function '_mm_extract_epi8' [-Wimplicit-function-declaration]
         B11[0] = _mm_extract_epi8(*L11, 0);
         ^
./include/snow3g_common.h: In function 'ClockLFSR_4':
./include/snow3g_common.h:196:9: warning: implicit declaration of function '_mm_insert_epi32' [-Wimplicit-function-declaration]
         T = _mm_insert_epi32(T, X2[0], 0);
         ^
./include/snow3g_common.h:196:11: error: incompatible types when assigning to type '__m128i' from type 'int'
         T = _mm_insert_epi32(T, X2[0], 0);
           ^
./include/snow3g_common.h:197:11: error: incompatible types when assigning to type '__m128i' from type 'int'
         T = _mm_insert_epi32(T, X2[1], 1);
           ^
./include/snow3g_common.h:198:11: error: incompatible types when assigning to type '__m128i' from type 'int'
         T = _mm_insert_epi32(T, X2[2], 2);
           ^
./include/snow3g_common.h:199:11: error: incompatible types when assigning to type '__m128i' from type 'int'
         T = _mm_insert_epi32(T, X2[3], 3);
           ^
In file included from ./include/snow3g_common.h:44:0,
                 from sse/snow3g_sse.c:42:
./include/snow3g_common.h: In function 'ClockFSM_4':
./include/snow3g_internal.h:344:27: error: incompatible types when assigning to type '__m128i' from type 'int'
                         w = _mm_insert_epi32(w, k, (n - 1));                   \
                           ^
./include/snow3g_common.h:302:9: note: in expansion of macro 'S1_S2_4'
         S1_S2_4(pCtx->FSM_X[2], pCtx->FSM_X[1], pCtx->FSM_X[0], K, L, 0);
         ^
./include/snow3g_internal.h:346:27: error: incompatible types when assigning to type '__m128i' from type 'int'
                         y = _mm_insert_epi32(y, ty, (n - 1));                  \
                           ^
./include/snow3g_common.h:302:9: note: in expansion of macro 'S1_S2_4'
         S1_S2_4(pCtx->FSM_X[2], pCtx->FSM_X[1], pCtx->FSM_X[0], K, L, 0);
         ^
./include/snow3g_internal.h:360:27: error: incompatible types when assigning to type '__m128i' from type 'int'
                         w = _mm_insert_epi32(w, k, n);                         \
                           ^
./include/snow3g_common.h:302:9: note: in expansion of macro 'S1_S2_4'
         S1_S2_4(pCtx->FSM_X[2], pCtx->FSM_X[1], pCtx->FSM_X[0], K, L, 0);
         ^
./include/snow3g_internal.h:362:27: error: incompatible types when assigning to type '__m128i' from type 'int'
                         y = _mm_insert_epi32(y, l, n);                         \
                           ^
./include/snow3g_common.h:302:9: note: in expansion of macro 'S1_S2_4'
         S1_S2_4(pCtx->FSM_X[2], pCtx->FSM_X[1], pCtx->FSM_X[0], K, L, 0);
         ^
./include/snow3g_internal.h:344:27: error: incompatible types when assigning to type '__m128i' from type 'int'
                         w = _mm_insert_epi32(w, k, (n - 1));                   \
                           ^
./include/snow3g_common.h:303:9: note: in expansion of macro 'S1_S2_4'
         S1_S2_4(pCtx->FSM_X[2], pCtx->FSM_X[1], pCtx->FSM_X[0], K, L, 1);
         ^
./include/snow3g_internal.h:346:27: error: incompatible types when assigning to type '__m128i' from type 'int'
                         y = _mm_insert_epi32(y, ty, (n - 1));                  \
                           ^
./include/snow3g_common.h:303:9: note: in expansion of macro 'S1_S2_4'
         S1_S2_4(pCtx->FSM_X[2], pCtx->FSM_X[1], pCtx->FSM_X[0], K, L, 1);
         ^
./include/snow3g_internal.h:360:27: error: incompatible types when assigning to type '__m128i' from type 'int'
                         w = _mm_insert_epi32(w, k, n);                         \
                           ^
./include/snow3g_common.h:303:9: note: in expansion of macro 'S1_S2_4'
         S1_S2_4(pCtx->FSM_X[2], pCtx->FSM_X[1], pCtx->FSM_X[0], K, L, 1);
         ^
./include/snow3g_internal.h:362:27: error: incompatible types when assigning to type '__m128i' from type 'int'
                         y = _mm_insert_epi32(y, l, n);                         \
                           ^
./include/snow3g_common.h:303:9: note: in expansion of macro 'S1_S2_4'
         S1_S2_4(pCtx->FSM_X[2], pCtx->FSM_X[1], pCtx->FSM_X[0], K, L, 1);
         ^
./include/snow3g_internal.h:344:27: error: incompatible types when assigning to type '__m128i' from type 'int'
                         w = _mm_insert_epi32(w, k, (n - 1));                   \
                           ^
./include/snow3g_common.h:304:9: note: in expansion of macro 'S1_S2_4'
         S1_S2_4(pCtx->FSM_X[2], pCtx->FSM_X[1], pCtx->FSM_X[0], K, L, 2);
         ^
./include/snow3g_internal.h:346:27: error: incompatible types when assigning to type '__m128i' from type 'int'
                         y = _mm_insert_epi32(y, ty, (n - 1));                  \
                           ^
./include/snow3g_common.h:304:9: note: in expansion of macro 'S1_S2_4'
         S1_S2_4(pCtx->FSM_X[2], pCtx->FSM_X[1], pCtx->FSM_X[0], K, L, 2);
         ^
./include/snow3g_internal.h:360:27: error: incompatible types when assigning to type '__m128i' from type 'int'
                         w = _mm_insert_epi32(w, k, n);                         \
                           ^
./include/snow3g_common.h:304:9: note: in expansion of macro 'S1_S2_4'
         S1_S2_4(pCtx->FSM_X[2], pCtx->FSM_X[1], pCtx->FSM_X[0], K, L, 2);
         ^
./include/snow3g_internal.h:362:27: error: incompatible types when assigning to type '__m128i' from type 'int'
                         y = _mm_insert_epi32(y, l, n);                         \
                           ^
./include/snow3g_common.h:304:9: note: in expansion of macro 'S1_S2_4'
         S1_S2_4(pCtx->FSM_X[2], pCtx->FSM_X[1], pCtx->FSM_X[0], K, L, 2);
         ^
./include/snow3g_internal.h:344:27: error: incompatible types when assigning to type '__m128i' from type 'int'
                         w = _mm_insert_epi32(w, k, (n - 1));                   \
                           ^
./include/snow3g_common.h:305:9: note: in expansion of macro 'S1_S2_4'
         S1_S2_4(pCtx->FSM_X[2], pCtx->FSM_X[1], pCtx->FSM_X[0], K, L, 3);
         ^
./include/snow3g_internal.h:346:27: error: incompatible types when assigning to type '__m128i' from type 'int'
                         y = _mm_insert_epi32(y, ty, (n - 1));                  \
                           ^
./include/snow3g_common.h:305:9: note: in expansion of macro 'S1_S2_4'
         S1_S2_4(pCtx->FSM_X[2], pCtx->FSM_X[1], pCtx->FSM_X[0], K, L, 3);
         ^
./include/snow3g_internal.h:360:27: error: incompatible types when assigning to type '__m128i' from type 'int'
                         w = _mm_insert_epi32(w, k, n);                         \
                           ^
./include/snow3g_common.h:305:9: note: in expansion of macro 'S1_S2_4'
         S1_S2_4(pCtx->FSM_X[2], pCtx->FSM_X[1], pCtx->FSM_X[0], K, L, 3);
         ^
./include/snow3g_internal.h:362:27: error: incompatible types when assigning to type '__m128i' from type 'int'
                         y = _mm_insert_epi32(y, l, n);                         \
                           ^
./include/snow3g_common.h:305:9: note: in expansion of macro 'S1_S2_4'
         S1_S2_4(pCtx->FSM_X[2], pCtx->FSM_X[1], pCtx->FSM_X[0], K, L, 3);
         ^
In file included from sse/snow3g_sse.c:42:0:
./include/snow3g_common.h: In function 'snow3gStateInitialize_4':
./include/snow3g_common.h:646:9: warning: implicit declaration of function '_mm_shuffle_epi8' [-Wimplicit-function-declaration]
         R = _mm_shuffle_epi8(R, *swapMask);
         ^
./include/snow3g_common.h:646:11: error: incompatible types when assigning to type '__m128i' from type 'int'
         R = _mm_shuffle_epi8(R, *swapMask);
           ^
./include/snow3g_common.h:647:11: error: incompatible types when assigning to type '__m128i' from type 'int'
         S = _mm_shuffle_epi8(S, *swapMask);
           ^
./include/snow3g_common.h:648:11: error: incompatible types when assigning to type '__m128i' from type 'int'
         T = _mm_shuffle_epi8(T, *swapMask);
           ^
./include/snow3g_common.h:649:11: error: incompatible types when assigning to type '__m128i' from type 'int'
         U = _mm_shuffle_epi8(U, *swapMask);
           ^
./include/snow3g_common.h: In function 'snow3gStateConvert_4':
./include/snow3g_common.h:1272:25: warning: implicit declaration of function '_mm_extract_epi32' [-Wimplicit-function-declaration]
                         T = _mm_extract_epi32(LFSR_X[(i + iLFSR_X) % 16], 0);
                         ^
./include/snow3g_common.h: In function 'snow3g_f8_4_buffer_sse':
./include/snow3g_common.h:1574:43: warning: implicit declaration of function '_mm_extract_epi64' [-Wimplicit-function-declaration]
                                           _mm_extract_epi64(L, 0));
                                           ^
make[2]: *** [obj/snow3g_sse.o] Error 1
`

## Possible Solution

I'm guessing a compiler deficiency, but I'm not finding a minimal compiler spec. Besides that, the SSE code just looks wrong (but, I'm way out of practice on SSE...)

## Steps to Reproduce

make clean
CONFIG: CONFIG_CRYPT=y
./configure
make

## Context (Environment including OS version, SPDK version, etc.)

CentOS 7.5 gcc 4.8.5 SPDK 20.04.1 (DPDK subproject)

**Comment by nufosmatic** [comment]
This is apparently in the included Intel Multi-Buffer library, and these tools meet the minimums:

5. Compilation
==============

Linux (64-bit only)
-------------------

Required tools:
- GNU make
- NASM version 2.13.03 (or newer)
- gcc (GCC) 4.8.3 (or newer)

----

$ nasm -v
NASM version 2.13.03 compiled on Jun  5 2018
$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/4.8.5/lto-wrapper
Target: x86_64-redhat-linux
Configured with: ../configure --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-bootstrap --enable-shared --enable-threads=posix --enable-checking=release --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-linker-hash-style=gnu --enable-languages=c,c++,objc,obj-c++,java,fortran,ada,go,lto --enable-plugin --enable-initfini-array --disable-libgcj --with-isl=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-x86_64-redhat-linux/isl-install --with-cloog=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-x86_64-redhat-linux/cloog-install --enable-gnu-indirect-function --with-tune=generic --with-arch_32=x86-64 --build=x86_64-redhat-linux
Thread model: posix
gcc version 4.8.5 20150623 (Red Hat 4.8.5-39) (GCC)

**Comment by peluse** [comment]
SPK is on ver .51 however DPDK supports up to ver .54.  No idea if it will help you or not but as you have the env already setup can you please try this? https://github.com/01org/intel-ipsec-mb/archive/v0.54.zip

**Comment by nufosmatic** [comment]
The CONFIG_CRYPT=y option requires at least SSE4.2 to build properly. However, this feature does not appear in the gcc by default on the box I'm using:

$ gcc -dM -E - < /dev/null | egrep SSE
#define __SSE2_MATH__ 1
#define __SSE_MATH__ 1
#define __SSE2__ 1
#define __SSE__ 1
$ gcc -msse4 -dM -E - < /dev/null | egrep SSE
#define __SSE4_1__ 1
#define __SSE4_2__ 1
#define __SSE2_MATH__ 1
#define __SSE_MATH__ 1
#define __SSE2__ 1
#define __SSSE3__ 1
#define __SSE__ 1
#define __SSE3__ 1

So the Intel Multi-Buffering will fail to build complaining about _mm_extract_epi32().

The workaround is to explicitly add the -msse4.2 to the MB build:

make -C intel-ipsec-mb all SHARED=n EXTRA_CFLAGS='-msse4.2 -fPIC' CC=cc

the original command line can be found by running make verbosely:

make V=1

It builds, but I still don't know that this works...

**Comment by nufosmatic** [comment]
I've determined that the CRYPT feature does not do what I need, so I'm not going to pursue installing and testing the alternative intel library (as I'm still not certain the offending feature is involved).

However, I did have a look at it, and it does look like it addresses the specific problem I found:

```
> # Set generic architectural optimizations
> OPT_SSE := -msse4.2 -maes
> OPT_AVX := -mavx -maes
> OPT_AVX2 := -mavx2 -maes
> OPT_AVX512 := -mavx2 -maes # -mavx512f is not available until gcc 4.9
> OPT_NOAESNI := -msse4.2 -mno-aes

```
